README.md
# moter-modem

Interface program to interact with RFM69 based modem boards using [CBOR](https://github.com/ekoeppen/ocaml-cbor)


debian/changelog
moter-modem (1.2.0-1) bionic; urgency=medium

  * Add SOCKS5 proxy support

 -- Eckhart Koeppen <eck@40hz.org>  Sat, 05 Jan 2019 07:25:41 +0200

moter-modem (1.1.0-1) bionic; urgency=medium

  * Change to JSON format for message payloads

 -- Eckhart Koeppen <eck@40hz.org>  Fri, 04 Jan 2019 14:46:52 +0200

moter-modem (1.0.2-1) bionic; urgency=medium

  * Fix incorrect parameter order

 -- Eckhart Koeppen <eck@40hz.org>  Wed, 02 Jan 2019 12:46:00 +0200

moter-modem (1.0.1-1) bionic; urgency=medium

  * Repackage with new files

 -- Eckhart Koeppen <eck@40hz.org>  Fri, 21 Dec 2018 05:25:11 +0000

moter-modem (1.0.0-1) bionic; urgency=low

  * Initial Release.

 -- Eckhart Koeppen <eck@40hz.org>  Tue, 18 Dec 2018 16:47:56 +0000


tests/cbor_test.ml
open Cmdliner

let major_test s =
  let major = Cbor.major_of_string s |> Cbor.string_of_major in
  Logs.info (fun m -> m "Major: %s" major)

let int_test s =
  match Cbor.int_of_string s with
  | Some (n, _) -> Logs.info (fun m -> m "Integer: %d" n)
  | None -> Logs.info (fun m -> m "Not an integer")

let int_2x_test s =
  let open Core.Option.Let_syntax in
  let%bind (n1, s) = Cbor.int_of_string s in
  let%map (n2, _) = Cbor.int_of_string s in
  Logs.info (fun m -> m "Integers: %d %d" n1 n2)

let float_test s =
  match Cbor.float_of_string s with
  | Some (n, _) -> Logs.info (fun m -> m "Float: %f" n)
  | None -> Logs.info (fun m -> m "No a float")

let tag_test s =
  match Cbor.tag_of_string s with
  | Some (n, _) -> Logs.info (fun m -> m "Tag: %d" n)
  | None -> Logs.info (fun m -> m "Not a tag")

let string_test s =
  match Cbor.byte_string_of_string s with
  | Some (str, _) -> Logs.info (fun m -> m "String: %s" str)
  | None -> Logs.info (fun m -> m "Not a string")

let test _ =
  major_test "\x01";
  major_test "\x21";
  major_test "\x41";
  major_test "\x61";
  major_test "\x81";
  major_test "\xa1";
  major_test "\xc1";
  major_test "\xe1";
  int_test "\x01";
  int_test "\x25";
  int_test "\xb1";
  let _ = int_2x_test "\x01\x02" in
  let _ = int_2x_test "\x18\x3f\x1a\x01\x01\x01\x01" in
  float_test "\xc4\x82\x21\x19\x6a\xb3";
  tag_test "\xc1";
  string_test "\x45Hello";
  ()

let setup_log style_renderer level =
  Fmt_tty.setup_std_outputs ?style_renderer ();
  Logs.set_level level;
  Logs.set_reporter (Logs_fmt.reporter ());
  ()

let logging =
  let env = Arg.env_var "CBOR_TEST_VERBOSITY" in
  Term.(const setup_log $ Fmt_cli.style_renderer () $ Logs_cli.level ~env ())

let cmd =
  let doc = "CBOR Test" in
  let exits = Term.default_exits in
  Term.(const test $ logging),
  Term.info "cbor_test" ~doc ~exits

let () = Term.(eval cmd |> exit)



tests/serial_test.ml
open Cmdliner

let dle = '\x10'
let stx = '\x02'
let etx = '\x03'

let test _ device =
  let b = Bytes.create 80 in
  let f = Unix.openfile device [Unix.O_RDWR] 0644 in
  Logs.info (fun m -> m "Device %s opened" device);
  let ic = Unix.in_channel_of_descr f in
  let oc = Unix.out_channel_of_descr f in
  output_string oc "Hello\r\n"; flush oc;
  let n = input ic b 0 80 in
  Logs.info (fun m -> m "Read %d bytes" n);
  let n = input ic b 0 80 in
  Logs.info (fun m -> m "Read %d bytes" n);
  Unix.close f

let rec wait_for_packet ic =
  let%lwt c1 = Lwt_io.read_char ic in
  let%lwt c2 = Lwt_io.read_char ic in
  if c1 != dle || c2 != stx
  then wait_for_packet ic
  else Lwt.return ()

let read_escaped_char ic =
  let%lwt c = Lwt_io.read_char ic in
  if c == dle then Lwt_io.read_char ic
  else Lwt.return c

let rec read_packet ic buffer =
  let%lwt c = read_escaped_char ic in
  if c != etx
  then (Buffer.add_char buffer c; read_packet ic buffer)
  else Lwt.return ()

let lwt_main device =
  let f = Unix.openfile device [Unix.O_RDWR] 0644 in
  Logs.info (fun m -> m "Device %s opened" device);
  let ic = Lwt_io.of_unix_fd ~mode:Lwt_io.input f in
  let oc = Unix.out_channel_of_descr f in
  let buffer = Buffer.create 16 in
  output_string oc "\x10\x02Hello, World!\x10\x03"; flush oc;
  let%lwt () = wait_for_packet ic in
  let%lwt () = read_packet ic buffer in
  Logs.info (fun m -> m "Read %s" (Buffer.contents buffer));
  Unix.close f;
  Lwt.return ()

let test_lwt _ device =
  Lwt_main.run (lwt_main device)

let device =
  let doc = "Device" in
  Arg.(required & pos 0 (some string) None & info [] ~docv:"DEVICE" ~doc)

let setup_log style_renderer level =
  Fmt_tty.setup_std_outputs ?style_renderer ();
  Logs.set_level level;
  Logs.set_reporter (Logs_fmt.reporter ());
  ()

let logging =
  let env = Arg.env_var "SERIAL_TEST_VERBOSITY" in
  Term.(const setup_log $ Fmt_cli.style_renderer () $ Logs_cli.level ~env ())

let cmd =
  let doc = "Serial Test" in
  let exits = Term.default_exits in
  Term.(const test_lwt $ logging $ device),
  Term.info "serial_test" ~doc ~exits

let () = Term.(eval cmd |> exit)


cbor.ml
type cbor_major_t =
  UNSIGNED_INT
  | NEGATIVE_INT
  | BYTE_STRING
  | UTF8_STRING
  | ARRAY
  | MAP
  | TAG
  | SIMPLE_OR_FLOAT
  | ERROR

let major_of_int = function
  | 0 -> UNSIGNED_INT
  | 1 -> NEGATIVE_INT
  | 2 -> BYTE_STRING
  | 3 -> UTF8_STRING
  | 4 -> ARRAY
  | 5 -> MAP
  | 6 -> TAG
  | 7 -> SIMPLE_OR_FLOAT
  | _ -> ERROR

let string_of_major = function
  | UNSIGNED_INT -> "unsigned int"
  | NEGATIVE_INT -> "negative int"
  | BYTE_STRING -> "byte string"
  | UTF8_STRING -> "utf8 string"
  | ARRAY -> "array"
  | MAP -> "map"
  | TAG -> "tag"
  | SIMPLE_OR_FLOAT -> "simple_or_float"
  | ERROR -> "error"

(* **** Low level decoding ************************************************* *)

let additional_info first =
  first land 0b00011111

let major_of_first first =
  major_of_int ((first land 0b11100000) lsr 5)

let decode_first s =
  let first = String.get s 0 |> Char.code in
  (major_of_first first, additional_info first)

let major_of_string s =
  String.get s 0 |> Char.code |> major_of_first

let decode_int additional s =
  match additional with
  | n when n < 24 -> (n, Core.String.drop_prefix s 1)
  | 24 -> (String.get s 1 |> Char.code, Core.String.drop_prefix s 2)
  | 25 ->
      ((String.get s 1 |> Char.code) lsl 8 +
       (String.get s 2 |> Char.code),
    Core.String.drop_prefix s 3)
  | 26 ->
      ((String.get s 1 |> Char.code) lsl 24 +
       (String.get s 2 |> Char.code) lsl 16 +
       (String.get s 3 |> Char.code) lsl 8 +
       (String.get s 4 |> Char.code), Core.String.drop_prefix s 5)
  | 27 -> raise (Failure "Not implemented")
  | _ -> raise (Failure "Incorrect additional information")

let decode_negative_int additional s =
  let (value, s) = decode_int additional s in
  (-1 - value, s)

let decode_byte_string additional s =
  let (length, s) = decode_int additional s in
  let str = Core.String.prefix s length in
  (str, Core.String.drop_prefix s length)

(* **** Decoding of CBOR elements ****************************************** *)

let int_of_string s =
  let (major, additional) = decode_first s in
  match major with
  | UNSIGNED_INT -> Some (decode_int additional s)
  | NEGATIVE_INT -> Some (decode_negative_int additional s)
  | _ -> None

let byte_string_of_string s =
  let (major, additional) = decode_first s in
  if major = BYTE_STRING
  then Some (decode_byte_string additional s)
  else None

let tag_of_string s =
  let (major, additional) = decode_first s in
  if major = TAG
  then Some (decode_int additional s)
  else None

let array_of_string s =
  let (major, additional) = decode_first s in
  if major = ARRAY
  then Some (decode_int additional s)
  else None

let map_of_string s =
  let (major, additional) = decode_first s in
  if major = MAP
  then Some (decode_int additional s)
  else None

let float_of_string s =
  let open Core.Option.Let_syntax in
  let (major, additional) = decode_first s in
  match major with
  | SIMPLE_OR_FLOAT -> None
  | TAG ->
      let tag, s = decode_int additional s in
      if tag == 4 then begin
        let%bind n, s = array_of_string s in
        if n == 2 then begin
          let%bind exponent, s = int_of_string s in
          let%bind mantissa, s = int_of_string s in
          Some (float_of_int mantissa *. 10.0 ** (float_of_int exponent), s)
        end
        else
          None
      end
      else
        None
  | _ -> None


conn.ml
open Rresult

type certs_t = {
  key : string;
  cert : string;
  ca : string;
}

let valid certs =
  certs.ca <> "" && certs.cert <> "" && certs.key <> ""

let get_proxy_setting proxy =
  let uri = Uri.of_string proxy in
  match Uri.host uri with
  | Some host -> begin match Uri.port uri with
    | Some port -> (host, port)
    | None -> (host, 1080)
    end
  | None -> ("", 0)

let connect_socket host port =
  let socket = Lwt_unix.socket Lwt_unix.PF_INET Lwt_unix.SOCK_STREAM 0 in
  let%lwt host_info = Lwt_unix.gethostbyname host in
  let server_address = host_info.Lwt_unix.h_addr_list.(0) in
  let%lwt () = Lwt_unix.connect socket (Lwt_unix.ADDR_INET (server_address, port)) in
  Logs.debug (fun m -> m "Socket connection to host %s" host);
  Lwt.return socket

let send_connect_request socket host port =
  let connect_str = R.get_ok (Socks_client.make_socks5_request (Connect {address = Domain_address host; port = port})) in
  Lwt_unix.write_string socket connect_str 0 (String.length connect_str)

let proxy_connect_socket proxy_host proxy_port host port =
  let%lwt socket = connect_socket proxy_host proxy_port in
  let auth_request = Socks_client.make_socks5_auth_request ~username_password:false in
  let auth_response = Bytes.make 2 '\x00' in
  let connect_response = Bytes.make 10 '\x00' in
  let%lwt _ = Lwt_unix.write_string socket auth_request 0 (String.length auth_request) in
  let%lwt _ = Lwt_unix.read socket auth_response 0 2 in
  let auth_method = Socks_client.parse_socks5_auth_response (Bytes.to_string auth_response) in
  begin match auth_method with
    | No_acceptable_methods -> Logs.err (fun m -> m "No acceptable auth methods")
    | _ -> Logs.debug (fun m -> m "Auth OK")
  end;
  let%lwt _ = send_connect_request socket host port in
  let%lwt _ = Lwt_unix.read socket connect_response 0 10 in
  let c = Socks_client.parse_socks5_response (Bytes.to_string connect_response) in
  begin match c with
    | Ok _ -> Logs.debug (fun m -> m "Connect request ok")
    | Error _ -> Logs.err (fun m -> m "Connect failed")
  end;
  Lwt.return socket

let of_socket socket authenticator certificate =
  Tls_lwt.Unix.client_of_fd Tls.Config.(client ~authenticator ~certificates:(`Single certificate) ~ciphers:Ciphers.supported ()) socket

let tls_connect ~host ~port ~proxy ~certs =
  let%lwt authenticator = X509_lwt.authenticator (`Ca_file certs.ca) in
  let%lwt certificate = X509_lwt.private_of_pems ~cert:certs.cert ~priv_key:certs.key in
  let (proxy_host, proxy_port) = get_proxy_setting proxy in
  if proxy_host <> "" then
    let () = Logs.info (fun m -> m "Connecting via proxy %s:%d" proxy_host proxy_port) in
    let%lwt socket = proxy_connect_socket proxy_host proxy_port host port in
    let%lwt tls = of_socket socket authenticator certificate in
    Logs.debug (fun m -> m "TLS handshake complete");
    let (ic, oc) = Tls_lwt.of_t tls in
    Lwt.return (ic, oc)
  else
    let%lwt (ic, oc) = Tls_lwt.connect_ext Tls.Config.(
      client ~authenticator ~certificates:(`Single certificate)
      ~ciphers:Ciphers.supported ()) (host, port) in
    Lwt.return (ic, oc)

let std_connect ~host ~port ~proxy =
  let (proxy_host, proxy_port) = get_proxy_setting proxy in
  let%lwt socket = (if proxy_host <> "" then
    let () = Logs.info (fun m -> m "Connecting via proxy %s:%d" proxy_host proxy_port) in
    proxy_connect_socket proxy_host proxy_port host port
  else
    connect_socket host port) in
  let oc = Lwt_io.of_fd ~mode:Lwt_io.output socket in
  let ic = Lwt_io.of_fd ~mode:Lwt_io.input socket in
  Lwt.return (ic, oc)

let connect ~host ~port ~proxy ~certs =
  Logs.debug (fun m -> m "Connecting to %s" host);
  let%lwt conn = (if (valid certs) then
      tls_connect ~host ~port ~proxy ~certs
    else
      std_connect ~host ~port ~proxy) in
  Logs.debug (fun m -> m "Connected to %s" host);
  Lwt.return conn


mqtt.ml
open Rresult
open Mqtt_lwt

type certs_t = {
  key : string;
  cert : string;
  ca : string;
}

let valid certs =
  certs.ca <> "" && certs.cert <> "" && certs.key <> ""

let get_proxy_setting proxy =
  let uri = Uri.of_string proxy in
  match Uri.host uri with
  | Some host -> begin match Uri.port uri with
    | Some port -> (host, port)
    | None -> (host, 1080)
    end
  | None -> ("", 0)

let connect_to_host host port =
  let socket = Lwt_unix.socket Lwt_unix.PF_INET Lwt_unix.SOCK_STREAM 0 in
  let%lwt host_info = Lwt_unix.gethostbyname host in
  let server_address = host_info.Lwt_unix.h_addr_list.(0) in
  let%lwt () = Lwt_unix.connect socket (Lwt_unix.ADDR_INET (server_address, port)) in
  Logs.debug (fun m -> m "Connected to host %s" host);
  Lwt.return socket

let send_connect_request socket host port =
  let connect_str = R.get_ok (Socks_client.make_socks5_request (Connect {address = Domain_address host; port = port})) in
  Lwt_unix.write_string socket connect_str 0 (String.length connect_str)

let connect proxy_host proxy_port host port =
  let%lwt socket = connect_to_host proxy_host proxy_port in
  let auth_request = Socks_client.make_socks5_auth_request ~username_password:false in
  let auth_response = Bytes.make 2 '\x00' in
  let connect_response = Bytes.make 10 '\x00' in
  let%lwt _ = Lwt_unix.write_string socket auth_request 0 (String.length auth_request) in
  let%lwt _ = Lwt_unix.read socket auth_response 0 2 in
  let auth_method = Socks_client.parse_socks5_auth_response (Bytes.to_string auth_response) in
  begin match auth_method with
    | No_acceptable_methods -> Logs.err (fun m -> m "No acceptable auth methods")
    | _ -> Logs.debug (fun m -> m "Auth OK")
  end;
  let%lwt _ = send_connect_request socket host port in
  let%lwt _ = Lwt_unix.read socket connect_response 0 10 in
  let c = Socks_client.parse_socks5_response (Bytes.to_string connect_response) in
  begin match c with
    | Ok _ -> Logs.debug (fun m -> m "Connect request ok")
    | Error _ -> Logs.err (fun m -> m "Connect failed")
  end;
  Lwt.return socket

let of_socket socket authenticator certificate =
  Tls_lwt.Unix.client_of_fd Tls.Config.(client ~authenticator ~certificates:(`Single certificate) ~ciphers:Ciphers.supported ()) socket

let tls_connect ~broker ~port ~proxy ~certs =
  let%lwt authenticator = X509_lwt.authenticator (`Ca_file certs.ca) in
  let%lwt certificate = X509_lwt.private_of_pems ~cert:certs.cert ~priv_key:certs.key in
  let (proxy_host, proxy_port) = get_proxy_setting proxy in
  if proxy_host <> "" then
    let () = Logs.info (fun m -> m "Connecting via proxy %s:%d" proxy_host proxy_port) in
    let%lwt socket = connect proxy_host proxy_port broker port in
    let%lwt tls = of_socket socket authenticator certificate in
    Logs.debug (fun m -> m "TLS handshake complete");
    let (ic, oc) = Tls_lwt.of_t tls in
    Lwt.return {ic; oc}
  else
    let%lwt (ic, oc) = Tls_lwt.connect_ext Tls.Config.(
      client ~authenticator ~certificates:(`Single certificate)
      ~ciphers:Ciphers.supported ()) (broker, port) in
    Lwt.return {ic; oc}

let std_connect ~broker ~port ~proxy =
  let (proxy_host, proxy_port) = get_proxy_setting proxy in
  let%lwt socket = (if proxy_host <> "" then
    connect proxy_host proxy_port broker port
  else
    Mqtt_lwt.connect ~host:broker ~port:port) in
  Logs.debug (fun m -> m "Connected to %s" broker);
  Lwt.return (Mqtt_lwt.of_socket socket)

let start_client ~id ~broker ~port ~proxy ~certs =
  Logs.debug (fun m -> m "Connecting to %s" broker);
  let%lwt conn = (if (valid certs) then
      tls_connect ~broker ~port ~proxy ~certs
    else
      std_connect ~broker ~port ~proxy) in
  Mqtt_lwt.mqtt_client conn ~opts:{default_conn_opts with client_id = id}

let sub topic client =
  Logs.debug (fun m -> m "Subscribing to topic %s" topic);
  Mqtt_lwt.subscribe ~topics:[topic] client.oc

let pub topic message client =
  Logs.debug (fun m -> m "Publishing %s to topic %s" message topic);
  Mqtt_lwt.publish ~topic:topic ~payload:message client.oc

let process client ~f =
  Mqtt_lwt.process_publish_pkt client f


serial.ml
let dle = '\x10'
let stx = '\x02'
let etx = '\x03'

let rec wait_for_packet ic =
  let%lwt c1 = Lwt_io.read_char ic in
  let%lwt c2 = Lwt_io.read_char ic in
  if c1 == dle && c2 == stx
  then Lwt.return ()
  else wait_for_packet ic

let read_escaped_char ic =
  let%lwt c = Lwt_io.read_char ic in
  (* Logs.debug (fun m -> m "%02x" (Char.code c)); *)
  if c == dle then begin
    let%lwt c = Lwt_io.read_char ic in
    (* Logs.debug (fun m -> m "%02x" (Char.code c)); *)
    if c != etx then Lwt.return (Some c)
    else Lwt.return None
  end
  else
    Lwt.return (Some c)

let rec read_packet ic buffer =
  let%lwt c = read_escaped_char ic in
  match c with
  | Some (c) -> (Buffer.add_char buffer c; read_packet ic buffer)
  | None -> Lwt.return ()

let write_packet oc buffer =
  let b = Buffer.create 64 in
  Buffer.add_char b dle; Buffer.add_char b stx;
  String.iter
    (fun c -> if c == dle then Buffer.add_char b dle; Buffer.add_char b c)
    buffer;
  Buffer.add_char b dle; Buffer.add_char b etx;
  Buffer.output_buffer oc b;
  flush oc

let open_device device =
  let f = Unix.openfile device [Unix.O_RDWR] 0644 in
  Logs.info (fun m -> m "Device %s opened" device);
  let state = (ExtUnixSpecific.Ioctl.tiocmget f) land (lnot 0x004) in
  let attr = Unix.tcgetattr f in
  let new_attr = {
    attr with
    c_icrnl = false;
    c_opost = false;
    c_icanon = false;
    c_isig = false;
    c_echo = false;
    c_echoe = false;
    c_echok = false;
    c_ixon = false;
    c_ixoff = false;
    c_ibaud = 115200;
    c_obaud = 115200;
  } in
  let () = Unix.tcsetattr f Unix.TCSANOW new_attr in
  let () = ExtUnixSpecific.Ioctl.tiocmset f state in
  let ic = Lwt_io.of_unix_fd ~mode:Lwt_io.input f in
  let oc = Unix.out_channel_of_descr f in
  (ic, oc)


modem.ml
open Cmdliner

type tag_t =
  | Sensor_reading_tag
  | Voltage_tag
  | Temperature_tag
  | Humidity_tag
  | Pressure_tag
  | Lux_tag
  | Uv_index_tag
  | Motion_tag
  | Sound_level_tag
  | Co2_tag
  | Test_packet_tag
  | Heartbeat_tag
  | Log_message_tag
  | Ping_tag
  | Register_value_tag
  | Error_message_tag

type response_t = (string * string) list option

let tag_of_int t =
  match t with
  | 6 -> Sensor_reading_tag
  | 7 -> Voltage_tag
  | 8 -> Temperature_tag
  | 9 -> Humidity_tag
  | 10 -> Pressure_tag
  | 11 -> Lux_tag
  | 12 -> Uv_index_tag
  | 13 -> Motion_tag
  | 14 -> Sound_level_tag
  | 15 -> Co2_tag
  | 64 -> Test_packet_tag
  | 65 -> Heartbeat_tag
  | 66 -> Log_message_tag
  | 67 -> Ping_tag
  | 68 -> Register_value_tag
  | 69 -> Error_message_tag
  | _ -> raise (Failure "Not a known tag")

let string_of_tag t =
  match t with
  | Sensor_reading_tag -> "Sensor reading"
  | Voltage_tag -> "Voltage"
  | Temperature_tag -> "Temperature"
  | Humidity_tag -> "Humidity"
  | Pressure_tag -> "Pressure"
  | Lux_tag -> "Lux"
  | Uv_index_tag -> "UV index"
  | Motion_tag -> "Motion"
  | Sound_level_tag -> "Sound level"
  | Co2_tag -> "CO2"
  | Test_packet_tag -> "Test packet"
  | Heartbeat_tag -> "Heartbeat"
  | Log_message_tag -> "Log message"
  | Ping_tag -> "Ping"
  | Register_value_tag -> "Register value"
  | Error_message_tag -> "Error message"

let to_success = function
  | None -> false
  | _ -> true

type channels_t = {
  ic : Lwt_io.input Lwt_io.channel;
  oc : out_channel;
  client : Mqtt_lwt.t;
  prefix : string;
}

let handle_heartbeat s =
  let open Core.Option.Let_syntax in
  match%bind Cbor.array_of_string s with
  | (2, s) ->
    let%bind node, s = Cbor.byte_string_of_string s in
    let%map seq, _s = Cbor.int_of_string s in
    Logs.info (fun m -> m "Heartbeat: %s %d" node seq);
    [
      (
        "/" ^ node,
        (Printf.sprintf "{\"heartbeat\":%d,\"ts\":%.0f}" seq (Unix.time ()))
      )
    ]
  | _ -> None

let handle_log_message s : response_t =
  let open Core.Option.Let_syntax in
  let%map message, _ = Cbor.byte_string_of_string s in
  Logs.info (fun m -> m "Log: %s" message);
  [
    (
      "/Modem",
      (Printf.sprintf "{\"log_message\":\"%s\",\"ts\":%.0f}" message (Unix.time ()))
    )
  ]

let handle_register_value s : response_t =
  let open Core.Option.Let_syntax in
  let%bind _n, s = Cbor.array_of_string s in
  let%bind register, s = Cbor.int_of_string s in
  let%map value, _ = Cbor.int_of_string s in
  Logs.info (fun m -> m "Register: %02x %02x" register value);
  [
    (
      "/Modem",
      (Printf.sprintf "{\"register\":{\"%d\":%d},\"ts\":%.0f}" register value (Unix.time ()))
    )
  ]

let handle_error_message s : response_t =
  let open Core.Option.Let_syntax in
  let%bind _n, s = Cbor.array_of_string s in
  let%bind file, s = Cbor.byte_string_of_string s in
  let%map line, _ = Cbor.int_of_string s in
  Logs.err (fun m -> m "Program error: %s %d" file line);
  [
    (
      "/Modem",
      (Printf.sprintf "{\"error\":\"%s:%d\",\"ts\":%.0f}" file line (Unix.time ()))
    )
  ]

let handle_ping s =
  let open Core.Option.Let_syntax in
  let%map node, _s = Cbor.byte_string_of_string s in
  Logs.info (fun m -> m "Ping: %s" node);
  [
    (
      "/Modem",
      (Printf.sprintf "{\"ping\":\"%s\",\"ts\":%.0f}" node (Unix.time ()))
    )
  ]

let handle_test_packet s =
  let open Core.Option.Let_syntax in
  let%map v, _s = Cbor.int_of_string s in
  [
    (
      "/Modem",
      (Printf.sprintf "{\"test\":\"%d\",\"ts\":%.0f}" v (Unix.time ()))
    )
  ]

let rec sensor_values_of_string n s values =
  if n > 1 then begin
    let open Core.Option.Let_syntax in
    let%bind t, s = Cbor.tag_of_string s in
    match Cbor.float_of_string s with
    | Some (v, s) ->
        let tag = tag_of_int t in
        Logs.debug (fun m -> m "%s: %f" (string_of_tag tag) v);
        sensor_values_of_string (n - 1) s ((tag, v) :: values)
    | _ -> None
  end
  else
    Some (values, s)

let string_of_values values =
  let s = Core.List.fold
    ~init:""
    ~f:(fun acc value ->
      acc ^ (Printf.sprintf "\"%s\":%f," (fst value |> string_of_tag) (snd value)))
    values in
  s ^ (Printf.sprintf "\"ts\":%.0f" (Unix.time ()))

let rec publish_messages m client prefix =
  match m with
  | hd :: tl ->
      let%lwt () = Mqtt.pub (prefix ^ (fst hd)) (snd hd) client in
      publish_messages tl client prefix
  | [] -> Lwt.return ()

let handle_sensor_reading s =
  let open Core.Option.Let_syntax in
  Logs.info (fun m -> m "Sensor data");
  let%bind n, s = Cbor.array_of_string s in
  let%bind node, s = Cbor.byte_string_of_string s in
  let%map values, _s = sensor_values_of_string n s [] in
  [
    (
      "/" ^ node,
      Printf.sprintf "{%s}" (string_of_values values)
    )
  ]

let handle_message s tag =
  let t = tag_of_int tag in
  Logs.debug (fun m -> m "Tag: %s" (string_of_tag t));
  match t with
    | Log_message_tag -> handle_log_message s
    | Heartbeat_tag -> handle_heartbeat s
    | Test_packet_tag -> handle_test_packet s
    | Ping_tag -> handle_ping s
    | Sensor_reading_tag -> handle_sensor_reading s
    | Register_value_tag -> handle_register_value s
    | Error_message_tag -> handle_error_message s
    | _ -> Logs.err (fun m -> m "Not a valid start tag %d" tag); None

let handle_packet s client prefix =
  Logs.debug (fun m -> m "Packet: ");
  match Cbor.tag_of_string s with
  | Some (tag, s) ->
      (match handle_message s tag with
      | Some response -> publish_messages response client prefix
      | None -> Lwt.return ())
  | None ->
      (Logs.info (fun m -> m "Message does not start with tag");
      Lwt.return ())

let rec modem_loop channels =
  let packet = Buffer.create 80 in
  let%lwt () = Serial.wait_for_packet channels.ic in
  let%lwt () = Serial.read_packet channels.ic packet in
  let%lwt _ok = handle_packet (Buffer.contents packet)
    channels.client channels.prefix in
  modem_loop channels

let display_topic channels _ topic payload id =
  Logs.info (fun m -> m "Topic: %s Payload: %s Msg_id: %d" topic payload id);
  let cmd = `Hex payload in
  let bin_cmd = Hex.to_string cmd in
  Serial.write_packet channels.oc bin_cmd;
  Lwt.return ()

let rec command_loop channels =
  let%lwt () = Mqtt.process channels.client ~f:(display_topic channels) in
  command_loop channels

let modem _logging ~device ~broker ~port ~proxy ~prefix ~certs =
  Logs.debug (fun m -> m "Start reporting to %s:%d/%s" broker port prefix);
  if (Conn.valid certs) then Logs.debug (fun m -> m "Certs: %s %s %s" certs.cert certs.key certs.ca);
  if (proxy <> "") then Logs.debug (fun m -> m "Proxy: %s" proxy);
  let ic, oc = Serial.open_device device in
  let id = Random.self_init (); Random.bits () |>
    Printf.sprintf "mqtt_lwt_%d" in
  let%lwt (mqtt_ic, mqtt_oc) = Conn.connect ~host:broker ~port ~proxy ~certs in
  let%lwt client = Mqtt_lwt.connect {oc = mqtt_oc; ic = mqtt_ic} ~opts:{Mqtt_lwt.default_conn_opts with client_id = id} in
  let%lwt () = Mqtt_lwt.subscribe ~topics:[prefix ^ "/Modem/Cmd"] client in
  let channels = {ic = ic; oc = oc; prefix = prefix; client = client} in
  Lwt.pick [
    Mqtt_lwt.process_publish_pkt client (display_topic channels);
    Mqtt_lwt.run client;
    modem_loop channels;
    command_loop channels
  ]

let lwt_wrapper logging device broker port proxy ca_file cert_file key_file prefix =
  let certs : Conn.certs_t = {
    cert = cert_file;
    key = key_file;
    ca = ca_file} in
  Lwt_main.run (modem logging ~device ~broker ~port ~proxy ~prefix ~certs)

let setup_log style_renderer level =
  Fmt_tty.setup_std_outputs ?style_renderer ();
  Logs.set_level level;
  Logs.set_reporter (Logs_fmt.reporter ());
  ()

let logging_arg =
  let env = Arg.env_var "MOTER_MODEM_VERBOSITY" in
  Term.(const setup_log $ Fmt_cli.style_renderer () $ Logs_cli.level ~env ())

let proxy_arg =
  let doc = "SOCKS5 proxy" in
  let env = Arg.env_var "MOTER_SOCKS_PROXY" in
  Arg.(value & opt string "" & info ["proxy"] ~env ~doc)

let device_arg =
  let doc = "Device" in
  let env = Arg.env_var "MOTER_DEVICE" in
  Arg.(value & opt string "/dev/ttyUSB0" & info ["d"; "device"] ~env ~doc)

let broker_arg =
  let doc = "MQTT broker" in
  Arg.(value & opt string "localhost" & info ["h"; "host"] ~doc)

let port_arg =
  let doc = "Port number" in
  Arg.(value & opt int 1883 & info ["p"; "port"] ~doc)

let prefix_arg =
  let doc = "MQTT prefix" in
  Arg.(value & opt string "" & info ["prefix"] ~doc)

let ca_file =
  let doc = "CA file" in
  Arg.(value & opt non_dir_file "" & info ["ca"] ~doc)

let cert_file =
  let doc = "Client certificate file" in
  Arg.(value & opt non_dir_file "" & info ["cert"] ~doc)

let key_file =
  let doc = "Client key" in
  Arg.(value & opt non_dir_file "" & info ["key"] ~doc)

let cmd =
  let doc = "MoteR Modem" in
  let exits = Term.default_exits in
  Term.(const lwt_wrapper $ logging_arg $ device_arg $
    broker_arg $ port_arg $ proxy_arg $
    ca_file $ cert_file $ key_file $
    prefix_arg),
  Term.info "moter-modem" ~doc ~exits

let () = Term.(eval cmd |> exit)
